app/src/main/java/org/feature/fox/coffee_counter/ui/CoreActivity.kt

```diff
@@ -41,5 +52,78 @@ class CoreActivity : ComponentActivity() {
                 }
             }
         }
+        mNfcAdapter = NfcAdapter.getDefaultAdapter(this)
+    }
+
+    override fun onResume() {
+        super.onResume()
+        if (mNfcAdapter != null) {
+            val options = Bundle()
+            // Work around for some broken Nfc firmware implementations that poll the card too fast
+            options.putInt(NfcAdapter.EXTRA_READER_PRESENCE_CHECK_DELAY, 250)
+
+            // Enable ReaderMode for all types of card and disable platform sounds
+            mNfcAdapter!!.enableReaderMode(
+                this,
+                this,
+                NfcAdapter.FLAG_READER_NFC_A or
+                        NfcAdapter.FLAG_READER_NFC_B or
+                        NfcAdapter.FLAG_READER_NFC_F or
+                        NfcAdapter.FLAG_READER_NFC_V or
+                        NfcAdapter.FLAG_READER_NFC_BARCODE or
+                        NfcAdapter.FLAG_READER_NO_PLATFORM_SOUNDS,
+                options
+            )
+        }
+    }
+
+    override fun onPause() {
+        super.onPause()
+        if (mNfcAdapter != null) mNfcAdapter!!.disableReaderMode(this)
+    }
+
+    override fun onTagDiscovered(tag: Tag) {
+        val mNdef: Ndef? = Ndef.get(tag)
+
+        if (mNdef == null){
+            runOnUiThread {
+                Toast.makeText(
+                    applicationContext,
+                    "something went wrong",
+                    Toast.LENGTH_SHORT
+                ).show()
+            }
+        }else{
+            try {
+                runOnUiThread {
+                    runBlocking {
+                        itemsViewModel.addStringItemToShoppingCart(readPayload(mNdef))
+                    }
+                }
+
+            } catch (e: FormatException) {
+                // if the NDEF Message to write is malformed
+            } catch (e: TagLostException) {
+                // Tag went out of range before operations were complete
+            } catch (e: IOException) {
+                // if there is an I/O failure, or the operation is cancelled
+            } finally {
```

Comment:
Should be annotated with TODO and if no logic is inside the catches it should be fine to just use the generic `e: Exception`